build-zones

```diff
@@ -0,0 +1,194 @@
+        return ipv4
+
+    @cached_property
+    def ipv6(self):
+        ipv6 = ipv4_to_ipv6(self.ipv4)
+        assert is_ocf_ip(ipv6), ipv6
+        return ipv6
+
+    @cached_property
+    def ldap_attrs(self):
+        with ldap_ocf() as c:
+            c.search(
+                OCF_LDAP_HOSTS,
+                '(cn={})'.format(self.hostname),
+                attributes=ldap3.ALL_ATTRIBUTES,
+            )
+            record, = c.response
+            return record['attributes']
+
+    @cached_property
+    def last_quad(self):
+        """Return the last quad from this host's IP address."""
+        return int(self.ipv4.exploded.split('.')[-1])
+
+
+    def forward_dns_records(self):
+        """Return forward DNS records."""
+        if self.ipv4:
+            yield '{self.hostname}\tIN A\t{self.ipv4}'.format(self=self)
+        if self.ipv6:
+            yield '{self.hostname}\tIN AAAA\t{self.ipv6}'.format(self=self)
+
+    def reverse_dns_records_v4(self):
+        """Return reverse IPv4 DNS records."""
+        if self.ipv4:
+            yield '{self.last_quad} IN PTR {self.hostname}.OCF.Berkeley.EDU.'.format(self=self)
+
+    def reverse_dns_records_v6(self):
+        """Return reverse IPv6 DNS records."""
+        last_quad_dotted = '.'.join(reversed(str(self.last_quad).zfill(4)))
+        # TODO: uppercase these properly like IPv4
+        if self.ipv6:
+            yield '{} IN PTR {self.hostname}.ocf.berkeley.edu.'.format(last_quad_dotted, self=self)
+
+    def __str__(self):
+        return '{self.hostname}\tIN A\t{self.ipv4}'.format(self=self)
+
+
+def hosts_from_ldap():
+    """Return set of hosts at the OCF."""
+    with ldap_ocf() as c:
+        c.search(OCF_LDAP_HOSTS, '(cn=*)', attributes=['cn'])
+        hosts = {
+            Host(record['attributes']['cn'][0])
+            for record in c.response
+        }
+    return hosts
+
+
+def forward_file(hosts, serial):
+    return open('templates/db.ocf.tmpl').read().format(
+        banner=BANNER,
+        serial=serial,
+        records='\n\n'.join(
+            '\n'.join(
+                ['; {}'.format(host.description)] +
+                list(host.forward_dns_records())
+            )
+            for host in sorted(hosts, key=attrgetter('ipv4'))
+        )
+    )
+
+
+def reverse_file_v4(hosts, serial):
+    return open('templates/db.ocf.in-addr.arpa.tmpl').read().format(
+        banner=BANNER,
+        serial=serial,
+        records='\n'.join(chain.from_iterable(
+            host.reverse_dns_records_v4()
+            for host in sorted(hosts, key=attrgetter('ipv4'))
+        ))
+    )
+
+
+def reverse_file_v6(hosts, serial):
+    return open('templates/db.ocf.ip6.arpa.tmpl').read().format(
+        banner=BANNER,
+        serial=serial,
+        records='\n'.join(chain.from_iterable(
+            host.reverse_dns_records_v6()
+            for host in sorted(hosts, key=attrgetter('ipv6'))
+        ))
+    )
+
+
+def generate_serial():
+    # TODO: this
```

Comment:
I agree that given to read from DNS or a zone file the file is less fickle.
Clearly both have chicken/egg problems but I don't really think that should
be an issue, we have existing records.

My existing method is at
https://github.com/nattofriends/suikoden/blob/master/suikoden/script.py .

On parsing an entire zone file vs. /etc/serial, not sure of the trade offs
between the two. Certainly it would be one other file to take care of and
remember that it exists.

On Thursday, February 25, 2016, Chris Kuehl <notifications@github.com>
wrote:

> In build-zones <https://github.com/ocf/dns/pull/2#discussion_r54201029>:
>
> > +    )
> > +
> > +
> > +def reverse_file_v6(hosts, serial):
> > +    return open('templates/db.ocf.ip6.arpa.tmpl').read().format(
> > +        banner=BANNER,
> > +        serial=serial,
> > +        records='\n'.join(chain.from_iterable(
> > +            host.reverse_dns_records_v6()
> > +            for host in sorted(hosts, key=attrgetter('ipv6'))
> > +        ))
> > +    )
> > +
> > +
> > +def generate_serial():
> > +    # TODO: this
>
> i'm a little concerned about reading from dns. there's a lot that can go
> wrong:
>
>    - network can be down
>    - dns can be down
>    - we might be running on the dns server (if we eventually move this to
>    puppet), possibly generating the zone file for the first time
>    - we can avoid some of this by reading from
>    - it doesn't work well with fast pushes (e.g. you notice a problem
>    right after pushing, you need to wait a couple minutes for jenkins to
>    deploy your broken records first, or manually set it)
>
> reading/writing a file would work possibly. something like this maybe?
>
> def next_serial():
>     """Return the next-greatest serial number from what's currently in use.    We don't want to assume networking is working, or that the DNS server is    up, or that the most recent changes have already propagated (e.g. they    might be half-way through a Jenkins pipeline).    So, instead we read it from a file we check into git.    """
>     date_as_str = time.strftime('%Y%m%d')
>     try:
>         serial = open('etc/serial').read().strip()
>         if serial.startswith(date_as_str) and len(serial) == 10:
>             return str(int(serial) + 1)
>     except IOError:
>         pass
>     return date_as_str + '00'
>
> (might still not work once it's in puppet, not sure)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ocf/dns/pull/2/files#r54201029>.
>